### ISSUE TITLE
XLA topk-decomposer crash in pmap(vmap(fn)) with top_k[0] in jnp.where + separate top_k[1]

### ISSUE THREAD

**Issue by jz-sa** [jax-ml/jax#36703]
## Bug

Combining `jax.lax.top_k(x, k)[0]` inside a `jnp.where` with a separate `jax.lax.top_k(y, k)[1]` call, inside `pmap(vmap(...))`, triggers an XLA HLO verifier error after the `topk-decomposer` pass.

**Error:**
```
INTERNAL: during context [hlo verifier]: Expected computation
compare_greater_than.1.1.clone.clone called from sort.2
to have 4 parameters, has 2

Failed after topk-decomposer
```

## Versions

- **Broken:** jax==0.9.0.1, jaxlib==0.9.0.1
- **Works:** jax==0.6.2, jaxlib==0.6.2 (latest stable)
- **Platform:** Linux, CUDA, 2x NVIDIA GPUs

## Minimal reproduction

```python
import jax
import jax.numpy as jnp

N = 50
K = 5
N_ITEMS = 20
N_GPUS = len(jax.devices("gpu"))
assert N_GPUS >= 2

padded = ((N_ITEMS + N_GPUS - 1) // N_GPUS) * N_GPUS
shard_indices = (jnp.arange(padded) % N_ITEMS).reshape(N_GPUS, -1)


def fn(i):
    x = jnp.arange(N, dtype=jnp.float32) * (1.0 + i * 0.01)

    # top_k values used inside jnp.where
    top_vals = jax.lax.top_k(x, K)[0]
    hard_sum = jnp.sum(top_vals)
    soft_sum = hard_sum * 0.99
    result = jnp.where(True, hard_sum, soft_sum)  # <-- trigger

    # top_k indices from separate call
    idx = jax.lax.top_k(x, K)[1]

    return result + jnp.float32(jnp.sum(idx))


# FAILS on 0.9.0.1, PASSES on 0.6.2
jax.pmap(jax.vmap(fn))(shard_indices)
```

## What works (any one condition removed)

| Variant | Result |
|---------|--------|
| `pmap(vmap(fn))` with `top_k[0]` in `where` + `top_k[1]` | **FAIL** |
| `vmap(fn)` — same fn, no pmap | PASS |
| `pmap(vmap(fn))` — remove `jnp.where` | PASS |
| `pmap(vmap(fn))` — remove `top_k[1]` | PASS |
| `pmap(vmap(fn))` — use only `top_k[0]` or only `top_k[1]` | PASS |

All three conditions are required to trigger the bug: (1) `pmap`, (2) `jnp.where` on a `top_k[0]`-derived value, (3) a separate `top_k[1]` call.

## Context

We hit this in a margin optimizer that uses `pmap(vmap(...))` for multi-GPU parallelism. The inner function computes multiple `soft_top_k_sum` calls (using `top_k[0]` with `jnp.where` for hard/soft switching) and one raw `top_k[1]` for mask creation. Downgrading to 0.6.2 resolved the issue.

**Comment by jz-sa** [comment]
Confirming this still reproduces on JAX 0.10.0 (latest stable) and on git HEAD `d211db6cdb04698ea6de27c081b88f956f878b4f`, tested 2026-04-16. Fresh run of the minimal repro (`fn_values_with_where_and_indices`) hits the same `Expected computation ... to have 4 parameters, has 2` / "Failed after topk-decomposer" error on 2× NVIDIA GPUs.

```
JAX 0.10.0, jaxlib 0.10.0, devices: [CudaDevice(id=0), CudaDevice(id=1)]

Test 1: pmap(vmap(top_k[0] in where + top_k[1]))
  FAIL: INTERNAL: during context [hlo verifier]: Expected computation compare_greater_than.1.1.clone.clone called from sort.2 to have 4 parameters, has 2

Failed after topk-decomposer
```

Ablations (Test 2–4) still pass — bug is specifically `pmap(vmap(fn))` with `top_k[0]` used inside `jnp.where` _and_ `top_k[1]` also called.

**Comment by mattjj** [comment]
We think https://github.com/openxla/xla/pull/41074 will fix this.

**Comment by mattjj** [comment]
https://github.com/openxla/xla/pull/41123 is a better fix: it removes a bad CSE behavior which was upstream of this topk stuff.

**Comment by mattjj** [comment]
https://github.com/openxla/xla/pull/41123 got merged, and so should be in tonight's nightly build.

**Comment by mattjj** [comment]
Can you check the fix is in [the nightly jaxlib](https://docs.jax.dev/en/latest/installation.html#jax-nightly-installation)?
